[APM] Syncs agent config settings to APM Fleet policies

ogupte opened the pull request:
Closes #95501.

Synchronizes all agent configuration settings with existing APM fleet policies whenever any user creates/updates any agent configuration via API endpoint `PUT /api/apm/settings/agent-configuration`.

Can be tested with apm-integration-testing `python3 ./scripts/compose.py start master --with-elastic-agent --apm-server-managed --all`. Update an agent confuration in APM UI, then inspect the response from `http://localhost:5601/api/fleet/package_policies` to verify that the agent configs are updated in the apm package policy object.

ogupte committed:
[APM] Syncs agent config settings to APM Fleet policies (#95501)

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/)
* Commit: c82917e0f2730afc2d3e0ea3243328908dbc024b
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Quick Commit Checks](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/execution/node/422/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/","number":"127553","commit":"c82917e0f2730afc2d3e0ea3243328908dbc024b"}]}
PIPELINE-->

ogupte committed:
fixes eslint issues

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/)
* Commit: b2bb6768d4bf51ee50d40f4322c740ac6e8ee794
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Quick Commit Checks](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/execution/node/400/log)

### History
* :broken_heart: [Build #127553](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/) failed c82917e0f2730afc2d3e0ea3243328908dbc024b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/","number":"127553","commit":"c82917e0f2730afc2d3e0ea3243328908dbc024b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/","number":"127582","commit":"b2bb6768d4bf51ee50d40f4322c740ac6e8ee794"}]}
PIPELINE-->

ogupte committed:
fixes malformed line comment

axw commented on x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts:
```diff
@@ -0,0 +1,100 @@
+    const configurations = await listConfigurations({ setup });
+    const agentConfigValue = configurations.map((configuration) => ({
+      service: configuration.service,
+      settings: configuration.settings,
```
```suggestion
      config: configuration.settings,
```

Here's how things look in apm-server currentlly: https://github.com/elastic/apm-server/blob/91645bde2d96a79b943c90253d62da3861ab2997/apm-server.yml#L275-L297

```
  #---------------------------- APM Server - Agent Configuration ----------------------------
  # Directly specify agent configuration. If `agent_config` is set, agent
  # configuration under `kibana` will be ignored.
  # An agent's incoming configuration request will be matched to an
  # agent_config with the following precedence:
  # - service.name and service.environment match an agent_config
  # - service.name matches an agent_config, service.environment == ""
  # - service.environment matches an agent_config, service.name == ""
  # - an agent_config without a name or environment set
  # An empty result is is returned if no matching result is found.
  # agent_config:
  # - service.name: ten_percent
  #   service.environment: production
  #   config:
  #     capture_body: off
  #     capture_body: true
  #     log_level: info
  #     recording: true
  #     transaction_sample_rate: 0.1
  # - service.name: frontend
  #   agent.name: rum-js
  #   config:
  #     transaction_sample_rate: 0.1
```

We could alternatively change the apm-server implementation if you/anyone feels strongly about calling this "settings".

ogupte replied:
I'm ok changing it to conform with apm-server's implementation. thanks!

axw commented on x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts:
```diff
@@ -0,0 +1,100 @@
+    const agentConfigValue = configurations.map((configuration) => ({
+      service: configuration.service,
+      settings: configuration.settings,
+    }));
```
```suggestion
      etag: configuration.etag,
      'agent.name': configuration.agent_name,
    }));
```

By passing the `etag` value down, apm-server will (later) be able to use it when recording which config has been applied by agents. The UI can then match this up to the config blocks it knows about.

We also need `agent.name` for restricting config for untrusted agents (RUM and Jaeger)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/)
* Commit: 9ff99d1d8e5bde0a6b963984e80030770ed681e2
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/9ff99d1d8e5bde0a6b963984e80030770ed681e2)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #127582](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/) failed b2bb6768d4bf51ee50d40f4322c740ac6e8ee794
* :broken_heart: [Build #127553](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/) failed c82917e0f2730afc2d3e0ea3243328908dbc024b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/","number":"127553","commit":"c82917e0f2730afc2d3e0ea3243328908dbc024b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/","number":"127582","commit":"b2bb6768d4bf51ee50d40f4322c740ac6e8ee794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/","number":"127585","commit":"9ff99d1d8e5bde0a6b963984e80030770ed681e2"}]}
PIPELINE-->

ogupte committed:
- consolidated logic that applies agent configurations to package policy objects
- update package policy agent_configs to include etag, agent.name, and change settings -> config

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/)
* Commit: c95249f7315c8de7701152138b97fe2cb3bfd56b
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/c95249f7315c8de7701152138b97fe2cb3bfd56b)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/090e0beb65e42b601b54b8f89afc701a3e1ec33d) | [after](https://github.com/elastic/kibana/commit/c95249f7315c8de7701152138b97fe2cb3bfd56b) | diff |
| --- | --- | --- | --- |
| `canvas` | 29 | 25 | -4 |
| `crossClusterReplication` | 8 | 6 | -2 |
| `fleet` | 22 | 20 | -2 |
| `globalSearch` | 4 | 2 | -2 |
| `indexManagement` | 12 | 7 | -5 |
| `infra` | 261 | 149 | -112 |
| `lens` | 67 | 45 | -22 |
| `licensing` | 18 | 15 | -3 |
| `lists` | 239 | 236 | -3 |
| `maps` | 286 | 208 | -78 |
| `ml` | 121 | 115 | -6 |
| `monitoring` | 109 | 56 | -53 |
| `securitySolution` | 386 | 342 | -44 |
| total |  |  | -336 |


</details>

### History
* :green_heart: [Build #127585](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/) succeeded 9ff99d1d8e5bde0a6b963984e80030770ed681e2
* :broken_heart: [Build #127582](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/) failed b2bb6768d4bf51ee50d40f4322c740ac6e8ee794
* :broken_heart: [Build #127553](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/) failed c82917e0f2730afc2d3e0ea3243328908dbc024b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/","number":"127553","commit":"c82917e0f2730afc2d3e0ea3243328908dbc024b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/","number":"127582","commit":"b2bb6768d4bf51ee50d40f4322c740ac6e8ee794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/","number":"127585","commit":"9ff99d1d8e5bde0a6b963984e80030770ed681e2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/","number":"127590","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"}]}
PIPELINE-->

cauemarcondes commented:
@ogupte could you give some instructions about how to test it, please?

ogupte committed:
Synchronizes agent configs whenever configuration is deleted.

cauemarcondes commented on x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts:
```diff
@@ -0,0 +1,122 @@
+  }
+  const fleetPluginStart = await plugins.fleet.start();
+
+  registerAgentConfigExternalCallback({
```
I think would be nice a comment explaining what are these callbacks.

stuartnelson3 commented:
I was running this branch locally, and I have a couple questions. Forgive my ignorance if this has been answered elsewhere before (:grimacing:):

1. I noticed my `agent_config` being rendered like this:
![image](https://user-images.githubusercontent.com/1398104/119853174-3e44a480-bf10-11eb-8424-3385917df860.png)
The `config` options are being rendered as strings, but they should be an int and a float. Is this expected?

2. The `agent_config` and `apm-server` blocks are separate. Does `agent_config` get merged into `apm-server` when it's pushed? And, should it be nested under `apm-server`?
![image](https://user-images.githubusercontent.com/1398104/119853715-bad78300-bf10-11eb-8469-9ea26a048c5d.png)

cauemarcondes commented on x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts:
```diff
@@ -0,0 +1,122 @@
+
+  registerAgentConfigExternalCallback({
+    fleetPluginStart,
+    callbackName: 'packagePolicyCreate',
```
I wonder if it is really necessary to have this callback `packagePolicyCreate`? When a policy is created it doesn't have an APM integration, so nothing is gonna happen, am I right? On the other hand, when you add APM integration to a policy it probably calls the `packagePolicyUpdate`.

cauemarcondes replied:
I made some tests and it looks like it's the other way around the `packagePolicyCreate` is called when APM integration is added to a policy. Hence I'm not sure when `packagePolicyUpdate` is invoked, would you know?

ogupte replied:
I'm not sure exactly when it would be implicitly invoked, but it could be explicitly invoked at any time outside of the APM plugin, so i think it's OK to keep it defined just in case.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/)
* Commit: c95249f7315c8de7701152138b97fe2cb3bfd56b
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/c95249f7315c8de7701152138b97fe2cb3bfd56b)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #127590](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127585](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/) succeeded 9ff99d1d8e5bde0a6b963984e80030770ed681e2
* :broken_heart: [Build #127582](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/) failed b2bb6768d4bf51ee50d40f4322c740ac6e8ee794
* :broken_heart: [Build #127553](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/) failed c82917e0f2730afc2d3e0ea3243328908dbc024b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/","number":"127553","commit":"c82917e0f2730afc2d3e0ea3243328908dbc024b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/","number":"127582","commit":"b2bb6768d4bf51ee50d40f4322c740ac6e8ee794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/","number":"127585","commit":"9ff99d1d8e5bde0a6b963984e80030770ed681e2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/","number":"127590","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/","number":"127674","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"}]}
PIPELINE-->

cauemarcondes commented on x-pack/plugins/apm/server/routes/settings/agent_configuration.ts:
```diff
@@ -95,10 +96,23 @@ const deleteAgentConfigurationRoute = createApmServerRoute({
+        setup,
+      });
+      logger.info(
+        `Saved latest agent settings to Fleet integration policy for APM.`
```
Since you're logging something, maybe would be good to have a specific message when it is deleting?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/)
* Commit: 200b23f7be35b0c7e960372954ee460a3cbf0082
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/200b23f7be35b0c7e960372954ee460a3cbf0082)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/090e0beb65e42b601b54b8f89afc701a3e1ec33d) | [after](https://github.com/elastic/kibana/commit/200b23f7be35b0c7e960372954ee460a3cbf0082) | diff |
| --- | --- | --- | --- |
| `canvas` | 29 | 25 | -4 |
| `crossClusterReplication` | 8 | 6 | -2 |
| `fleet` | 22 | 20 | -2 |
| `globalSearch` | 4 | 2 | -2 |
| `indexManagement` | 12 | 7 | -5 |
| `infra` | 261 | 149 | -112 |
| `lens` | 67 | 45 | -22 |
| `licensing` | 18 | 15 | -3 |
| `lists` | 239 | 236 | -3 |
| `maps` | 286 | 208 | -78 |
| `ml` | 121 | 115 | -6 |
| `monitoring` | 109 | 56 | -53 |
| `securitySolution` | 386 | 342 | -44 |
| total |  |  | -336 |


</details>

### History
* :green_heart: [Build #127674](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127590](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127585](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/) succeeded 9ff99d1d8e5bde0a6b963984e80030770ed681e2
* :broken_heart: [Build #127582](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/) failed b2bb6768d4bf51ee50d40f4322c740ac6e8ee794
* :broken_heart: [Build #127553](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/) failed c82917e0f2730afc2d3e0ea3243328908dbc024b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127553/","number":"127553","commit":"c82917e0f2730afc2d3e0ea3243328908dbc024b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/","number":"127582","commit":"b2bb6768d4bf51ee50d40f4322c740ac6e8ee794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/","number":"127585","commit":"9ff99d1d8e5bde0a6b963984e80030770ed681e2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/","number":"127590","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/","number":"127674","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/","number":"127699","commit":"200b23f7be35b0c7e960372954ee460a3cbf0082"}]}
PIPELINE-->

cauemarcondes reviewed (approved):
LGTM - Nice work @ogupte

ogupte commented:
> I was running this branch locally, and I have a couple questions. Forgive my ignorance if this has been answered elsewhere before (😬):
> 
>     1. I noticed my `agent_config` being rendered like this:
>        ![image](https://user-images.githubusercontent.com/1398104/119853174-3e44a480-bf10-11eb-8424-3385917df860.png)
>        The `config` options are being rendered as strings, but they should be an int and a float. Is this expected?
> 
>     2. The `agent_config` and `apm-server` blocks are separate. Does `agent_config` get merged into `apm-server` when it's pushed? And, should it be nested under `apm-server`?
>        ![image](https://user-images.githubusercontent.com/1398104/119853715-bad78300-bf10-11eb-8469-9ea26a048c5d.png)
@stuartnelson3, thanks for checking this out and testing it.
1. config options are intentionally strings, sort of like how ENV variables are always strings. This was something we decided when we implemented agent configurations in the first place.
2. While it's technically possible to combine Agent Configs and APM server configs into the same object, they configurations for different purposes, and it the APM server's responsibility to deliver agent configurations to the appropriate agent. If folks working on apm-server preferred them combined, i'll be happy to accommodate. @axw

ogupte committed:
PR feedback

ogupte committed:
Merge branch 'master' into apm-95501-agent-config-fleet-sync

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/)
* Commit: 49d660a1e2bd1845fdcda4ec1e7fb157354faffb
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/49d660a1e2bd1845fdcda4ec1e7fb157354faffb)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/be001f2aa689a2a6271a7d4c747f3ca18bb09be6) | [after](https://github.com/elastic/kibana/commit/49d660a1e2bd1845fdcda4ec1e7fb157354faffb) | diff |
| --- | --- | --- | --- |
| `stackAlerts` | 101 | 95 | -6 |


</details>

### History
* :green_heart: [Build #127699](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/) succeeded 200b23f7be35b0c7e960372954ee460a3cbf0082
* :green_heart: [Build #127674](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127590](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127585](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/) succeeded 9ff99d1d8e5bde0a6b963984e80030770ed681e2
* :broken_heart: [Build #127582](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/) failed b2bb6768d4bf51ee50d40f4322c740ac6e8ee794

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127582/","number":"127582","commit":"b2bb6768d4bf51ee50d40f4322c740ac6e8ee794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/","number":"127585","commit":"9ff99d1d8e5bde0a6b963984e80030770ed681e2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/","number":"127590","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/","number":"127674","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/","number":"127699","commit":"200b23f7be35b0c7e960372954ee460a3cbf0082"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/","number":"127766","commit":"49d660a1e2bd1845fdcda4ec1e7fb157354faffb"}]}
PIPELINE-->

axw commented:
> config options are intentionally strings, sort of like how ENV variables are always strings. This was something we decided when we implemented agent configurations in the first place.

:+1: sorry @stuartnelson3, I missed that we had numbers in our config in the `apm-server.yml` example. I just checked the code and it's a `map[string]string` so I think we just need to update the example.

> While it's technically possible to combine Agent Configs and APM server configs into the same object, they configurations for different purposes, and it the APM server's responsibility to deliver agent configurations to the appropriate agent. If folks working on apm-server preferred them combined, i'll be happy to accommodate. @axw

I would prefer that everything specific to APM is nested under the `apm-server` key, so `apm-server.agent_configs` in this case.

stuartnelson3 commented:
this ended up being correctly sent to `apm-server`, but I suppose it'd be nice to have it rendered within the apm-server config block:

```diff
diff --git a/x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts b/x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts
index bfc9d360..8ff34d96 100644
--- a/x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts
+++ b/x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts
@@ -106,7 +106,7 @@ export function getPackagePolicyWithAgentConfigurations(
       {
         ...firstInput,
         config: {
-          agent_config: {
+          'apm-server.agent_config': {
             value: agentConfigurations.map((configuration) => ({
               service: configuration.service,
               config: configuration.settings,
```
in the UI:
![image](https://user-images.githubusercontent.com/1398104/120301843-92afa180-c2cd-11eb-903a-ecc491d74a87.png)

ogupte commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'master' into apm-95501-agent-config-fleet-sync

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/)
* Commit: a25497ec3fcbe2011828fbe79930a94a227469af
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/a25497ec3fcbe2011828fbe79930a94a227469af)
* [Documentation Changes](https://kibana_100744.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/9734d1cfe8baff1879a2240456159c9b2033c63d) | [after](https://github.com/elastic/kibana/commit/a25497ec3fcbe2011828fbe79930a94a227469af) | diff |
| --- | --- | --- | --- |
| `canvas` | 29 | 25 | -4 |
| `crossClusterReplication` | 8 | 6 | -2 |
| `fleet` | 22 | 20 | -2 |
| `globalSearch` | 4 | 2 | -2 |
| `indexManagement` | 12 | 7 | -5 |
| `infra` | 256 | 146 | -110 |
| `lens` | 67 | 45 | -22 |
| `licensing` | 18 | 15 | -3 |
| `lists` | 239 | 236 | -3 |
| `maps` | 286 | 208 | -78 |
| `ml` | 121 | 115 | -6 |
| `monitoring` | 109 | 56 | -53 |
| `securitySolution` | 390 | 346 | -44 |
| `stackAlerts` | 101 | 95 | -6 |
| total |  |  | -340 |


</details>

### History
* :green_heart: [Build #127766](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/) succeeded 49d660a1e2bd1845fdcda4ec1e7fb157354faffb
* :green_heart: [Build #127699](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/) succeeded 200b23f7be35b0c7e960372954ee460a3cbf0082
* :green_heart: [Build #127674](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127590](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127585](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/) succeeded 9ff99d1d8e5bde0a6b963984e80030770ed681e2

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127585/","number":"127585","commit":"9ff99d1d8e5bde0a6b963984e80030770ed681e2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/","number":"127590","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/","number":"127674","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/","number":"127699","commit":"200b23f7be35b0c7e960372954ee460a3cbf0082"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/","number":"127766","commit":"49d660a1e2bd1845fdcda4ec1e7fb157354faffb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/","number":"128405","commit":"a25497ec3fcbe2011828fbe79930a94a227469af"}]}
PIPELINE-->

ogupte committed:
nest agent_config within `apm-server` in the package policy input

ogupte commented:
> this ended up being correctly sent to `apm-server`, but I suppose it'd be nice to have it rendered within the apm-server config block:
> ...

@stuartnelson3 can you confirm that this works: https://github.com/elastic/kibana/pull/100744/files#diff-ab7e62a4c711e2133e13004dec4ecfedd8bb1bf46da87eed88f61bd743b2ffd4R118-R128

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/)
* Commit: 6129f4ff1e06e358193539710c5a2f29ae8a9fa1
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/6129f4ff1e06e358193539710c5a2f29ae8a9fa1)
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)
* [Documentation Changes](https://kibana_100744.docs-preview.app.elstc.co/diff)

### Failed CI Steps
* [Execute xpack-kibana-ciGroup8](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/execution/node/721/log)
* [Execute kibana-ciGroup9](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/execution/node/845/log)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/9734d1cfe8baff1879a2240456159c9b2033c63d) | [after](https://github.com/elastic/kibana/commit/6129f4ff1e06e358193539710c5a2f29ae8a9fa1) | diff |
| --- | --- | --- | --- |
| `canvas` | 29 | 25 | -4 |
| `crossClusterReplication` | 8 | 6 | -2 |
| `fleet` | 22 | 20 | -2 |
| `globalSearch` | 4 | 2 | -2 |
| `indexManagement` | 12 | 7 | -5 |
| `infra` | 256 | 146 | -110 |
| `lens` | 67 | 45 | -22 |
| `licensing` | 18 | 15 | -3 |
| `lists` | 239 | 236 | -3 |
| `maps` | 286 | 208 | -78 |
| `ml` | 121 | 115 | -6 |
| `monitoring` | 109 | 56 | -53 |
| `securitySolution` | 390 | 346 | -44 |
| `stackAlerts` | 101 | 95 | -6 |
| total |  |  | -340 |


</details>

### History
* :green_heart: [Build #128405](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/) succeeded a25497ec3fcbe2011828fbe79930a94a227469af
* :green_heart: [Build #127766](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/) succeeded 49d660a1e2bd1845fdcda4ec1e7fb157354faffb
* :green_heart: [Build #127699](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/) succeeded 200b23f7be35b0c7e960372954ee460a3cbf0082
* :green_heart: [Build #127674](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b
* :green_heart: [Build #127590](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127590/","number":"127590","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/","number":"127674","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/","number":"127699","commit":"200b23f7be35b0c7e960372954ee460a3cbf0082"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/","number":"127766","commit":"49d660a1e2bd1845fdcda4ec1e7fb157354faffb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/","number":"128405","commit":"a25497ec3fcbe2011828fbe79930a94a227469af"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/","number":"128571","commit":"6129f4ff1e06e358193539710c5a2f29ae8a9fa1"}]}
PIPELINE-->

ogupte commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'master' into apm-95501-agent-config-fleet-sync

stuartnelson3 commented:
I started from a fresh ES + new build of kibana off this branch, and when I tried to add the APM server integration I got this error:

![image](https://user-images.githubusercontent.com/1398104/120492700-95d48b80-c3ba-11eb-8fb9-5463c1702af5.png)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/)
* Commit: 62379ad465ab1b1d99de9b843501a4f1f1a4e129
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/62379ad465ab1b1d99de9b843501a4f1f1a4e129)
* [Documentation Changes](https://kibana_100744.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #128571](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/) failed 6129f4ff1e06e358193539710c5a2f29ae8a9fa1
* :green_heart: [Build #128405](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/) succeeded a25497ec3fcbe2011828fbe79930a94a227469af
* :green_heart: [Build #127766](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/) succeeded 49d660a1e2bd1845fdcda4ec1e7fb157354faffb
* :green_heart: [Build #127699](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/) succeeded 200b23f7be35b0c7e960372954ee460a3cbf0082
* :green_heart: [Build #127674](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/) succeeded c95249f7315c8de7701152138b97fe2cb3bfd56b

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127674/","number":"127674","commit":"c95249f7315c8de7701152138b97fe2cb3bfd56b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/","number":"127699","commit":"200b23f7be35b0c7e960372954ee460a3cbf0082"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/","number":"127766","commit":"49d660a1e2bd1845fdcda4ec1e7fb157354faffb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/","number":"128405","commit":"a25497ec3fcbe2011828fbe79930a94a227469af"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/","number":"128571","commit":"6129f4ff1e06e358193539710c5a2f29ae8a9fa1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/","number":"128639","commit":"62379ad465ab1b1d99de9b843501a4f1f1a4e129"}]}
PIPELINE-->

ogupte committed:
nests agent_config under the requried 'value' property of config['apm-server']
in order to pass validation checks

ogupte commented:
@stuartnelson3 in order to fix the validation error I had to define a property `value` for `config['apm-server']` within the package policy. Overall the structure looks like this: `packagePolicy.inputs[0].config['apm-server'].value.agent_config = Array<{service, etrag, config: {...}}>`. I also see there's a `compiled_input` object in the input, that contains `apm-server` (`packagePolicy.inputs[0].compiled_input['apm-server'] = { api_key, secret_token, host, ... }`), but i assume we want to continue to use `config`. Can you confirm if this schema works for apm-server?

![Screen Shot 2021-06-02 at 4 40 24 PM](https://user-images.githubusercontent.com/1967266/120549917-324e5c00-c3c2-11eb-8fd7-5bb1d40501b4.png)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128767/)
* Commit: 2d522ce39705a882e658daf500ce8b61bba938e7
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/2d522ce39705a882e658daf500ce8b61bba938e7)
* [Documentation Changes](https://kibana_100744.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #128639](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/) succeeded 62379ad465ab1b1d99de9b843501a4f1f1a4e129
* :broken_heart: [Build #128571](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/) failed 6129f4ff1e06e358193539710c5a2f29ae8a9fa1
* :green_heart: [Build #128405](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/) succeeded a25497ec3fcbe2011828fbe79930a94a227469af
* :green_heart: [Build #127766](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/) succeeded 49d660a1e2bd1845fdcda4ec1e7fb157354faffb
* :green_heart: [Build #127699](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/) succeeded 200b23f7be35b0c7e960372954ee460a3cbf0082

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127699/","number":"127699","commit":"200b23f7be35b0c7e960372954ee460a3cbf0082"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/","number":"127766","commit":"49d660a1e2bd1845fdcda4ec1e7fb157354faffb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/","number":"128405","commit":"a25497ec3fcbe2011828fbe79930a94a227469af"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/","number":"128571","commit":"6129f4ff1e06e358193539710c5a2f29ae8a9fa1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/","number":"128639","commit":"62379ad465ab1b1d99de9b843501a4f1f1a4e129"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128767/","number":"128767","commit":"2d522ce39705a882e658daf500ce8b61bba938e7"}]}
PIPELINE-->

ogupte committed:
- externalizes getApmPackagePolicies for reusability
- parallelizes operations for improved performance

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128815/)
* Commit: 6793d6289e56943f2e103268ad2b3adcb46e04e1
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/6793d6289e56943f2e103268ad2b3adcb46e04e1)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128815/execution/node/646/log)
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128815/execution/node/861/log)

### History
* :green_heart: [Build #128767](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128767/) succeeded 2d522ce39705a882e658daf500ce8b61bba938e7
* :green_heart: [Build #128639](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/) succeeded 62379ad465ab1b1d99de9b843501a4f1f1a4e129
* :broken_heart: [Build #128571](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/) failed 6129f4ff1e06e358193539710c5a2f29ae8a9fa1
* :green_heart: [Build #128405](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/) succeeded a25497ec3fcbe2011828fbe79930a94a227469af
* :green_heart: [Build #127766](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/) succeeded 49d660a1e2bd1845fdcda4ec1e7fb157354faffb

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127766/","number":"127766","commit":"49d660a1e2bd1845fdcda4ec1e7fb157354faffb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/","number":"128405","commit":"a25497ec3fcbe2011828fbe79930a94a227469af"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/","number":"128571","commit":"6129f4ff1e06e358193539710c5a2f29ae8a9fa1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/","number":"128639","commit":"62379ad465ab1b1d99de9b843501a4f1f1a4e129"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128767/","number":"128767","commit":"2d522ce39705a882e658daf500ce8b61bba938e7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128815/","number":"128815","commit":"6793d6289e56943f2e103268ad2b3adcb46e04e1"}]}
PIPELINE-->

stuartnelson3 commented:
`config` is correct, cool. I pulled in the changes and no longer get the validation error when creating the integration, but after adding an agent config, I'm not seeing it rendered in the UI:
![image](https://user-images.githubusercontent.com/1398104/120602445-51450080-c44b-11eb-9a6e-77d70783713e.png)

ogupte commented:
> `config` is correct, cool. I pulled in the changes and no longer get the validation error when creating the integration, but after adding an agent config, I'm not seeing it rendered in the UI:
@stuartnelson3 It looks like the fleet kibana code which generates that yaml spreads in the contents of `config` followed by the contents of `compiled_input`, and since they both have properties named `apm-server` the latter overwrites the former: https://github.com/elastic/kibana/blob/master/x-pack/plugins/fleet/common/services/package_policies_to_agent_inputs.ts#L34-L38

is apm-server consuming these agent configs with the raw policy object, or only through this output yaml?

stuartnelson3 commented:
@ogupte I'm not sure, I've been doing manual testing by copying+pasting the output yaml. @axw had surmised this to possibly be the case -- maybe he also knows if the raw policy object is being pushed or not?

cauemarcondes commented on x-pack/plugins/apm/server/lib/fleet/register_fleet_policy_callbacks.ts:
```diff
@@ -0,0 +1,134 @@
+    inputs: [
+      {
+        ...firstInput,
+        config: {
+          [APM_SERVER]: {
```
@ogupte Don't you need to add the others `config` here to avoid overwriting the entire config object?
```suggestion
        config: {
          ...firstInput.config,
          [APM_SERVER]: {
```

axw commented:
> is apm-server consuming these agent configs with the raw policy object, or only through this output yaml?

@ogupte apm-server will receiver the combined output yaml. Maybe flatten the config string, so it's added like `apm-server.agent_config` rather than as a nested object structure.

ogupte commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'master' into apm-95501-agent-config-fleet-sync

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/129903/)
* Commit: aaf52635aae0604209ec58c2e7f1321b0a13ab42
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100744/aaf52635aae0604209ec58c2e7f1321b0a13ab42)
* [Documentation Changes](https://kibana_100744.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #128815](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128815/) failed 6793d6289e56943f2e103268ad2b3adcb46e04e1
* :green_heart: [Build #128767](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128767/) succeeded 2d522ce39705a882e658daf500ce8b61bba938e7
* :green_heart: [Build #128639](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/) succeeded 62379ad465ab1b1d99de9b843501a4f1f1a4e129
* :broken_heart: [Build #128571](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/) failed 6129f4ff1e06e358193539710c5a2f29ae8a9fa1
* :green_heart: [Build #128405](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/) succeeded a25497ec3fcbe2011828fbe79930a94a227469af

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128405/","number":"128405","commit":"a25497ec3fcbe2011828fbe79930a94a227469af"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128571/","number":"128571","commit":"6129f4ff1e06e358193539710c5a2f29ae8a9fa1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128639/","number":"128639","commit":"62379ad465ab1b1d99de9b843501a4f1f1a4e129"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128767/","number":"128767","commit":"2d522ce39705a882e658daf500ce8b61bba938e7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128815/","number":"128815","commit":"6793d6289e56943f2e103268ad2b3adcb46e04e1"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/129903/","number":"129903","commit":"aaf52635aae0604209ec58c2e7f1321b0a13ab42"}]}
PIPELINE-->
